### DISCUSSION TITLE
JVM hangs on stop / dump commands

### DISCUSSION THREAD

**Discussion by mtakeshi1** [async-profiler/async-profiler#439] [Q&A] [answered]
Hello everyone.

Has anyone experienced an issue where sometimes when the agent sends the stop or dump commands, the command never finishes and the jvm seems to hang - or at least it looks like it. The jvm gets irresponsive to jcmd, jstack and other commands (maybe because the thread that handles attach events is stuck?) and the application threads also seems to either get stuck or get really slow (and CPU usage drops correspondly). The process gets in a state that i have to kill -9 and restart. 

It only happens on our test server (did not try in the production servers yet) and its doesn't happen always (maybe like 25% of the time?) but it looks like it happens more often when the server is busy. 

I've added a bunch of log messages to try to find the cause and so far, I've got two cases: 
- one that the thread trying to stop the profiler gets stuck in the spinlock but never breaks out of it, as if an event is being handled forever. Further logs on Profiler::recordSample showed a thread locking on that and getting stuck right after calling getJavaTraceAsync
- the other case, while cleaning up the threads, it gets stuck right after closing (munmap) and unlocking the events page

I'm not knowledgeble in jvm internals (or c++ for that matter) so I don't know how else to debug it but to put a bunch of Log::info (I've also added a Mutex in there just because with so many log messages, the log was getting weird).
How do you guys usually debug issues like that?

Thanks a bunch,
takeshi

ps: I'm running on the latest jdk16 from Azul.

**Comment by apangin** [comment, marked as answer]
Thank you for the report.
Can you tell the exact command you use to run and to stop profiling?
What is the OS version?
Which GC is used?

Does the problem happen when you stop profiling without dumping (i.e. `stop` command without output file)?
Please also try the latest profiler version from master, as there were some bugfixes related to JVM hanging (#422, #415).

Whenever you see the problem next time, try [`gstack`](https://linux.die.net/man/1/gstack) - it will dump native stack traces of the JVM process.

**Reply by mtakeshi1** [reply]
Hi Andre

profiler.sh start -i 1000  -f /tmp/results2.jfr -t <proc_name>

I was running with a higher interval, but with 1us it seems to hang more frequently (it could be my own bias of couse)

The OS is an old CentOS (I forgot to check it in my original post sorry about that)
CentOS Linux release 7.5.1804 (Core) 

(I had to compile with stdc++ statically linked in order to add the log statements)

runnig a pretty old kernel:
Linux arke-titan-homolog 3.10.0-862.el7.x86_64

It happens both on G1 and ZGC.

will try again soon-ish (later today) and see what gstack tells me (thanks for the hint). Can I use the file generated by the perf-map-agent (the java agent that spits out the mapping file to match with perf stack traces) to try to make sense or are the addresses not related?

by the way, great tool and I fully expect it to be something that I did wrong 
Thanks a bunch,
takeshi

**Reply by mtakeshi1** [reply]
Ran again into the issue but sadly gstack did not give me any clues =(

All (or most) of the application threads are waiting on safepoints, gc threads are waiting, many threads are waiting on Unsafe_Park after calling (os::javaTimeNanos()). Strace (which I started running before the profiler.sh stop command) also did not give anything to work with.

The issue doesn't seem to happen if I run with wall profiler, which is what I will be doing from now on.

Thanks for the help.

**Reply by apangin** [reply]
Well, if the issue reproduces often in your environment, it will be nice to find out the reason and fix the bug, if any.

- Can you share gstack output? In particular, what is AttachListener thread is doing? Are there profiler functions anywhere on the stack?
- Does JVM hang when using other profiler dump option, e.g. flamegraph?
- What about `-e itimer` mode?
- What profiling interval `-i` do you specify?
- Does the process consume CPU, or all threads are idle?

**Reply by mtakeshi1** [reply]
I'm attaching the gstack output. 
[stack.log](https://github.com/jvm-profiling-tools/async-profiler/files/6618405/stack.log)

I had mentioned before that the application threads were on safepoints, but this stack doesn't show it so I was wrong (the previous stack trace did show lot of application threads on safepoints, but not on this last one).

The attach thread, I believe is this one and maybe it looks like its waiting to accept a new connection

Thread 11 (Thread 0x7f2f7e9ec700 (LWP 2067)):
#0  0x00007f30c9be6a7d in accept () from /lib64/libpthread.so.0
#1  0x00007f30c84302ef in LinuxAttachListener::dequeue() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#2  0x00007f30c8430437 in AttachListener::dequeue() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#3  0x00007f30c842f396 in attach_listener_thread_entry(JavaThread*, Thread*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#4  0x00007f30c8eebc4c in JavaThread::thread_main_inner() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#5  0x00007f30c8ef04df in Thread::call_run() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#6  0x00007f30c8c5ca77 in thread_native_entry(Thread*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#7  0x00007f30c9bdfe25 in start_thread () from /lib64/libpthread.so.0
#8  0x00007f30c9705bad in clone () from /lib64/libc.so.6

So far I've only tried with the start / stop pair, dump and collect and they all seem to hang. I'm gonna try with flamegraph and -e itimer. As for the profiling interval, I've seen it hang with 10ms but it happens more often with 1us (I have not tried to go further down)

The running process does consume CPU but much less than before issueing the stop command. The application also gets irresponsive to regular signals (SIGTRAP, etc) after the stop command and must be killed with -9

**Reply by apangin** [reply]
Hmm. This particular gstack looks normal. Are you sure it's taken in the state when the JVM is hung?
All threads are idle; I don't know which one may consume CPU. You may probably find a busy PID with `top -H` and them match it with LWP from gstack output.

Also, there is no sign of the profiler, and Attach Thread is also idle waiting for incoming connection.
Does the problem happen *before* `stop` command finishes, or *after*? According to the stack traces, it is definitely not in the middle of stopping profiler.

**Reply by mtakeshi1** [reply]
Yes this stack is from when the process is hang.

I'm gonna try to get a "better" stack later but when it happens, the stop command never finished (from the profiler.sh perspective). For some reason, I couldn't reproduce the freeze today, which is even stranger (even with -e cpu and with a high frequency).

Thanks for the help,
takeshi

**Reply by mtakeshi1** [reply]
Well after some days trying (and failing) to reproduce the problem, I will chalk this one as user error. Not sure what I was doing wrong but its working great now. 
Sorry to waste your time and thanks for the attention

**Reply by apangin** [reply]
No problem. Feel free to come back if you encounter something suspicious again.

**Comment by mtakeshi1** [comment]
my attempt to debug have introduced a hidden deadlock, so I'm back to square one =(
It may happen in real life I think, but I'm not too sure. The problematic stack trace looks like this:


Thread 28 (Thread 0x7f5fb77f7700 (LWP 22722)):
#0  0x00007f610155b89c in __lll_lock_wait_private () from /lib64/libc.so.6
#1  0x00007f610156530d in _L_lock_58 () from /lib64/libc.so.6
#2  0x00007f610156520f in __fprintf_chk () from /lib64/libc.so.6
#3  0x00007f5fb49ee3f0 in fprintf (__fmt=0x7f5fb4a9afc1 "%s\n", __stream=0x7f5fec044720) at /usr/include/x86_64-linux-gnu/bits/stdio2.h:98
#4  Log::log (level=level@entry=0x7f5fb4a9db0a "[INFO] ", msg=msg@entry=0x7f5fb4a9b260 "[%#010x] acquired lock_index %d", args=args@entry=0x7f5fb77f3e50) at src/log.cpp:62
#5  0x00007f5fb49ee0a7 in Log::info (msg=msg@entry=0x7f5fb4a9b260 "[%#010x] acquired lock_index %d") at src/log.cpp:69
#6  0x00007f5fb49c0286 in Profiler::recordSample (this=0x7f5fb4ce13c0 <Profiler::_instance>, ucontext=ucontext@entry=0x7f5fb77f4000, counter=783582, event_type=event_type@entry=0, event=even
t@entry=0x7f5fb77f3fc0) at src/profiler.cpp:598
#7  0x00007f5fb49de688 in PerfEvents::signalHandler (signo=<optimized out>, siginfo=0x7f5fb77f4130, ucontext=0x7f5fb77f4000) at src/perfEvents_linux.cpp:633
#8  <signal handler called>
#9  0x00007f61015651fd in __fprintf_chk () from /lib64/libc.so.6
#10 0x00007f5fb49ee3f0 in fprintf (__fmt=0x7f5fb4a9afc1 "%s\n", __stream=0x7f5fec044720) at /usr/include/x86_64-linux-gnu/bits/stdio2.h:98
#11 Log::log (level=level@entry=0x7f5fb4a9db0a "[INFO] ", msg=msg@entry=0x7f5fb4a9cc10 "waiting to __sync_bool_compare_and_swap on thread %d", args=args@entry=0x7f5fb77f5210) at src/log.cpp:
62
#12 0x00007f5fb49ee0a7 in Log::info (msg=msg@entry=0x7f5fb4a9cc10 "waiting to __sync_bool_compare_and_swap on thread %d") at src/log.cpp:69
#13 0x00007f5fb49de26f in PerfEvents::destroyForThread (tid=tid@entry=14644) at src/perfEvents_linux.cpp:593
#14 0x00007f5fb49de407 in destroyForThread (tid=14644) at src/perfEvents_linux.cpp:768
#15 PerfEvents::stop (this=<optimized out>) at src/perfEvents_linux.cpp:769
#16 0x00007f5fb49c2c47 in Profiler::stop (this=this@entry=0x7f5fb4ce13c0 <Profiler::_instance>) at src/profiler.cpp:1079
#17 0x00007f5fb49c6bac in Profiler::runInternal (this=0x7f5fb4ce13c0 <Profiler::_instance>, args=..., out=...) at src/profiler.cpp:1385
#18 0x00007f5fb49c775f in Profiler::run (this=0x7f5fb4ce13c0 <Profiler::_instance>, args=...) at src/profiler.cpp:1396
#19 0x00007f5fb49d2fed in Agent_OnAttach (vm=0x7f61013c6200 <main_vm>, options=<optimized out>, reserved=<optimized out>) at src/vmEntry.cpp:303
#20 0x00007f61008952e7 in JvmtiExport::load_agent_library(char const*, char const*, char const*, outputStream*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#21 0x00007f6100276cdd in load_agent(AttachOperation*, outputStream*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#22 0x00007f61002774a1 in attach_listener_thread_entry(JavaThread*, Thread*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#23 0x00007f6100d33c4c in JavaThread::thread_main_inner() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#24 0x00007f6100d384df in Thread::call_run() () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#25 0x00007f6100aa4a77 in thread_native_entry(Thread*) () from /opt/zulu16.28.11-ca-jdk16.0.0-linux_x64/lib/server/libjvm.so
#26 0x00007f6101a27e25 in start_thread () from /lib64/libpthread.so.0
#27 0x00007f610154dbad in clone () from /lib64/libc.so.6


it looks like the thread that is stopping the profiling (and writing to the file, hence holding the stream mutex) is also handling some signal (I don't even know how that is possible) and trying to write some log. I'm gonna remove all of the logging and try to reproduce the hangup. Thanks for the help!
